### PR TITLE
Gruppe 3 review

### PR DESCRIPTION
**Pull request by erendogan51**

### CODE REVIEW

File: .github/workflows/test_workflow.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+          distribution: 'adopt'
+          java-version: '17'
+      - name: Grant Permissions to gradlew
+        run: chmod +x gradlew
+      - name: Validate Gradle wrapper
+        uses: gradle/wrapper-validation-action@e6e38bacfdf1a337459f332974bb2327a31aaf4b
+      - name: Build with Gradle
+        uses: gradle/gradle-build-action@67421db6bd0bf253fb4bd25b31ebb98943c375e1
+        with:
+          arguments: build
```
Comment (erendogan51) [inline .github/workflows/test_workflow.yml:25]:
```suggestion
          arguments: test
```

File: src/main/java/at/ac/fhcampuswien/asd/entity/models/User.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    Long id;
+    final Long VALID_SESSION_TIME = 120 * 1000L;
+    String username;
+    String firstName;
+    String lastName;
+    int failedLoginCounter = 0;
+    Long lockedUntil = null;
+    Long sessionValidUntil;
+    UUID session;
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/entity/models/User.java:30]:
```suggestion
   private  String username;
   private String firstName;
   private String lastName;
   private int failedLoginCounter = 0;
   private Long lockedUntil = null;
   private Long sessionValidUntil;
   private UUID session;
```

File: src/main/java/at/ac/fhcampuswien/asd/helper/Hashing.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.KeySpec;
+import java.util.ArrayList;
+
+public class Hashing {
+
+    public static ArrayList<byte[]> generateHash(String password) {
+
+        ArrayList<byte[]> list = new ArrayList<>();
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/helper/Hashing.java:15]: List is useless, return byte[] directly

File: src/main/java/at/ac/fhcampuswien/asd/rest/controller/DeletionRestController.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import javax.validation.Valid;
+import java.util.UUID;
+
+@RestController
+@CrossOrigin
+@RequiredArgsConstructor
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+public class DeletionRestController {
+
+    UserRestService userRestService;
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/controller/DeletionRestController.java:35]:
```suggestion
    private UserRestService userRestService;
```

File: src/main/java/at/ac/fhcampuswien/asd/rest/controller/DeletionRestController.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                    ),
+                    @ApiResponse(description = "User don`t exists", responseCode = "404", content = @Content)
+            }
+    )
+    @DeleteMapping("/users/{username}")
+    public ResponseEntity<String> remove(@PathVariable String username,
+                                         @RequestBody @Valid DeleteUserRequest deleteUserRequest,
+                                         HttpSession session) throws UserNotFoundException, InvalidSessionException, InvalidPasswordException {
+
+        userRestService.removeUserByUsername(username, deleteUserRequest.getCurrentPassword(), (UUID) session.getAttribute(sessionIdName));
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/controller/DeletionRestController.java:62]: `(UUID) session.getAttribute(sessionIdName)` this is dangerous, typecast in service

File: src/main/java/at/ac/fhcampuswien/asd/rest/controller/PasswordChangeRestController.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                            description = "OK",
+                            responseCode = "200"
+                    ),
+                    @ApiResponse(description = "User not found", responseCode = "404", content = @Content),
+                    @ApiResponse(description = "Invalid session", responseCode = "401", content = @Content),
+                    @ApiResponse(description = "Invalid password", responseCode = "401", content = @Content),
+            }
+    )
+    public ResponseEntity<Object> changePassword(@PathVariable String username, @Valid @RequestBody InboundUserChangePasswordDto inboundUserChangePasswordDto, HttpSession session) throws InvalidSessionException, UserNotFoundException, InvalidPasswordException {
+        userService.changePassword(username, inboundUserChangePasswordDto, (UUID) session.getAttribute(sessionIdName));
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/controller/PasswordChangeRestController.java:59]: `(UUID) session.getAttribute(sessionIdName)` See above

File: src/main/java/at/ac/fhcampuswien/asd/rest/controller/RegistrationRestController.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    UserRestService userRestService;
+
+    /**
+     * Registers the user in the database.
+     *
+     * @param inboundUserRegistrationDto Specifies information to create the user account.
+     * @return Returns HTTP Created.
+     * @throws UserAlreadyExistsException In case the username is already taken.
+     */
+    @Operation(summary = "Creates a user in the database.", tags = {"Users"}, responses = {@ApiResponse(description = "Created", responseCode = "201", content = @Content(mediaType = "application/json", schema = @Schema(implementation = OutboundUserRegistrationDto.class))), @ApiResponse(description = "User already Exists", responseCode = "409", content = @Content)})
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/controller/RegistrationRestController.java:39]: User vertical line distribution not horizontal, line too long

File: src/main/java/at/ac/fhcampuswien/asd/rest/controller/SessionRestController.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * @param inboundUserLoginDto Specified username and password.
+     * @param session             HTTP Session object
+     * @return Returns OK.
+     * @throws UserLockedException     In case the user is locked.
+     * @throws AuthenticationException In case the user or password are not correct
+     */
+
+    @PostMapping("/login")
+    @Operation(summary = "Login a user.", tags = {"Users"}, responses = {@ApiResponse(description = "OK", responseCode = "200"), @ApiResponse(description = "User locked", responseCode = "401", content = @Content), @ApiResponse(description = "User or password not correct", responseCode = "401", content = @Content)})
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/controller/SessionRestController.java:44]: Line too long, see above

File: src/main/java/at/ac/fhcampuswien/asd/rest/controller/SessionRestController.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    /**
+     * @param username Specifies the user to be logged out.
+     * @param session  Specifies the session to close.
+     * @return Returns OK.
+     * @throws InvalidSessionException In case the session does not exist for the specified user.
+     * @throws UserNotFoundException   In case the user does not exist.
+     */
+
+    @PostMapping("/{username}/logout")
+    @Operation(summary = "Logs out a user.", tags = {"Users"}, responses = {@ApiResponse(description = "OK", responseCode = "200"), @ApiResponse(description = "User not found", responseCode = "404", content = @Content), @ApiResponse(description = "Invalid session", responseCode = "401", content = @Content)})
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/controller/SessionRestController.java:62]: Line too long, see above

File: src/main/java/at/ac/fhcampuswien/asd/rest/controller/SessionRestController.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    public ResponseEntity<Object> logout(@PathVariable String username, HttpSession session) throws InvalidSessionException, UserNotFoundException {
+
+        userService.logout(username, (UUID) session.getAttribute(sessionIdName));
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/controller/SessionRestController.java:65]: `(UUID) session.getAttribute(sessionIdName)` see above

File: src/main/java/at/ac/fhcampuswien/asd/rest/service/UserRestService.java [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    }
+
+    /**
+     * Checks the users' existence.
+     *
+     * @param username The username of the user to check.
+     * @return Returns the User.
+     * @throws UserNotFoundException In case the user is not found in the database.
+     */
+    private User checkUserExistence(String username) throws UserNotFoundException {
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/service/UserRestService.java:120]: Boolean expected with this method name

File: src/main/java/at/ac/fhcampuswien/asd/rest/service/UserRestService.java [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+        }
+    }
+
+    /**
+     * Checks the locked status.
+     *
+     * @param user The user for which to check the locked status.
+     * @throws UserLockedException In case the user is locked.
+     */
+    private void checkLockedStatus(User user) throws UserLockedException {
```
Comment (PeclBoi) [inline src/main/java/at/ac/fhcampuswien/asd/rest/service/UserRestService.java:151]: Method name, see above

**Review by PeclBoi — CHANGES_REQUESTED** [review]
Smaller suggestions, all in all looks good.

**Pull request closed without merging**